File: packages/astro-engine-python/src/astro_engine/planet_observation.py
"""Live planet observation facts for one night.

Normative procedure: contracts/procedures/planet-observation.md. This computes
night-scoped provider facts; `targets.planet_recommendation` consumes the result
deterministically.

The portable model is the production one, ported from
`LowPrecisionPlanetAstronomyProvider`:

* Schlyter low-precision orbital elements with the `JD - 2451543.5` day number;
* one eccentric-anomaly correction term, not an iterated Kepler solve;
* Earth's elements supply the geocentric Sun vector, which is added to the
  planet's heliocentric vector to get the geocentric one;
* geometric altitude/azimuth with no refraction, parallax or light-time;
* solar elongation as the angle between the geocentric planet and Sun vectors;
* samples by repeated addition of the cadence from `night_start - 7200` while
  the instant stays at or before `night_end + 3600`. The lead endpoint is always
  sampled; the trailing endpoint only when the cadence lands on it, and there is
  no explicit interval-end sample.

Operation ordering follows the Swift source so binary64 results agree.
"""

from __future__ import annotations

import math
import re
from datetime import datetime, timedelta, timezone
from typing import Any, Mapping, NamedTuple

from astro_engine.deep_sky_observation import exceeds_sample_cap
from astro_engine.errors import SampleCapError, ValidationError

CAPABILITY_ID = "astronomy.planet_observation"

DEFAULT_SAMPLE_INTERVAL = 15 * 60.0
LEAD_SECONDS = 2 * 3600.0
TRAIL_SECONDS = 60 * 60.0

# Astronomy-namespace instant window, shared with sun_events/moon_*/planet.
EARLIEST_EPOCH_SECONDS = 946_684_800.0     # 2000-01-01T00:00:00Z
LATEST_EPOCH_SECONDS = 2_524_607_999.0     # 2049-12-31T23:59:59Z

# Bounded live night interval; the same 26 hours the namespace already uses.
MAX_SPAN_SECONDS = 26 * 3600.0

# One day of one-minute cadence over the sampling span. Production asks for one
# astronomical night at 900 s — at most ~70 samples.
MAX_SAMPLE_COUNT = 1_440

# The production user-facing candidates. `earth` is in the model only to supply
# the geocentric Sun vector and is never a target.
SUPPORTED_TARGET_IDS = ("venus", "mars", "jupiter", "saturn")

# Swift `Date` counts seconds from 2001-01-01T00:00:00Z; matching the scale keeps
# sample arithmetic bit-identical, as in targets.deep_sky_windows.
_REFERENCE_EPOCH = 978_307_200.0
_EPOCH = datetime(1970, 1, 1, tzinfo=timezone.utc)
_PATTERN = re.compile(r"[0-9]{4}-[0-9]{2}-[0-9]{2}T[0-9]{2}:[0-9]{2}:[0-9]{2}Z", re.ASCII)


class Sample(NamedTuple):
    time: float
    """Swift reference-epoch seconds."""
    altitude: float
    azimuth: float
    solar_elongation: float


def _radians(degrees: float) -> float:
    return degrees * math.pi / 180


def _degrees(radians: float) -> float:
    return radians * 180 / math.pi


def _normalized_degrees(degrees: float) -> float:
    return math.fmod(math.fmod(degrees, 360.0) + 360.0, 360.0)


def _normalized_radians(radians: float) -> float:
    two_pi = 2 * math.pi
    return math.fmod(math.fmod(radians, two_pi) + two_pi, two_pi)


def _elements(body: str, d: float) -> tuple[float, float, float, float, float, float]:
    """`(node, inclination, argument_of_perihelion, a, e, mean_anomaly)`."""
    if body == "earth":
        return (0.0, 0.0, 282.9404 + 4.70935e-5 * d, 1.0,
                0.016709 - 1.151e-9 * d, 356.0470 + 0.9856002585 * d)
    if body == "venus":
        return (76.6799 + 2.46590e-5 * d, 3.3946 + 2.75e-8 * d, 54.8910 + 1.38374e-5 * d,
                0.723330, 0.006773 - 1.302e-9 * d, 48.0052 + 1.6021302244 * d)
    if body == "mars":
        return (49.5574 + 2.11081e-5 * d, 1.8497 - 1.78e-8 * d, 286.5016 + 2.92961e-5 * d,
                1.523688, 0.093405 + 2.516e-9 * d, 18.6021 + 0.5240207766 * d)
    if body == "jupiter":
        return (100.4542 + 2.76854e-5 * d, 1.3030 - 1.557e-7 * d, 273.8777 + 1.64505e-5 * d,
                5.20256, 0.048498 + 4.469e-9 * d, 19.8950 + 0.0830853001 * d)
    if body == "saturn":
        return (113.6634 + 2.38980e-5 * d, 2.4886 - 1.081e-7 * d, 339.3939 + 2.97661e-5 * d,
                9.55475, 0.055546 - 9.499e-9 * d, 316.9670 + 0.0334442282 * d)
    raise _invalid()


def _heliocentric(elements: tuple[float, float, float, float, float, float]
                  ) -> tuple[float, float, float]:
    node, inclination, argument_of_perihelion, semi_major_axis, eccentricity, mean_anomaly = elements
    mean_anomaly_radians = _radians(_normalized_degrees(mean_anomaly))
    eccentric_anomaly = mean_anomaly_radians + eccentricity * math.sin(mean_anomaly_radians) * (
        1 + eccentricity * math.cos(mean_anomaly_radians)
    )

    xv = semi_major_axis * (math.cos(eccentric_anomaly) - eccentricity)
    yv = semi_major_axis * math.sqrt(1 - eccentricity * eccentricity) * math.sin(eccentric_anomaly)
    true_anomaly = math.atan2(yv, xv)
    radius = math.sqrt(xv * xv + yv * yv)

    node_radians = _radians(node)
    inclination_radians = _radians(inclination)
    argument = true_anomaly + _radians(argument_of_perihelion)

    x = radius * (math.cos(node_radians) * math.cos(argument)
                  - math.sin(node_radians) * math.sin(argument) * math.cos(inclination_radians))
    y = radius * (math.sin(node_radians) * math.cos(argument)
                  + math.cos(node_radians) * math.sin(argument) * math.cos(inclination_radians))
    z = radius * math.sin(argument) * math.sin(inclination_radians)
    return x, y, z


def _angular_separation(first: tuple[float, float, float],
                        second: tuple[float, float, float]) -> float:
    dot = first[0] * second[0] + first[1] * second[1] + first[2] * second[2]
    first_magnitude = math.sqrt(first[0] * first[0] + first[1] * first[1] + first[2] * first[2])
    second_magnitude = math.sqrt(second[0] * second[0] + second[1] * second[1] + second[2] * second[2])
    if not (first_magnitude > 0 and second_magnitude > 0):
        return 0.0
    cosine = min(max(dot / (first_magnitude * second_magnitude), -1), 1)
    return _degrees(math.acos(cosine))


def position(body: str, latitude: float, longitude: float, reference_seconds: float) -> Sample:
    """Geocentric horizontal coordinates and solar elongation at one instant."""
    jd = (reference_seconds + _REFERENCE_EPOCH) / 86_400 + 2_440_587.5
    # Schlyter's day-number convention: d = JD - 2451543.5, not the J2000.0 epoch.
    d = jd - 2_451_543.5
    sun_geocentric = _heliocentric(_elements("earth", d))
    planet = _heliocentric(_elements(body, d))

    x = planet[0] + sun_geocentric[0]
    y = planet[1] + sun_geocentric[1]
    z = planet[2] + sun_geocentric[2]
    obliquity = _radians(23.4393 - 3.563e-7 * d)

    equatorial_x = x
    equatorial_y = y * math.cos(obliquity) - z * math.sin(obliquity)
    equatorial_z = y * math.sin(obliquity) + z * math.cos(obliquity)
    right_ascension = math.atan2(equatorial_y, equatorial_x)
    declination = math.atan2(
        equatorial_z, math.sqrt(equatorial_x * equatorial_x + equatorial_y * equatorial_y))

    local_sidereal_time = _radians(_normalized_degrees(
        280.460_618_37 + 360.985_647_366_29 * (jd - 2_451_545.0) + longitude
    ))
    hour_angle = _normalized_radians(local_sidereal_time - right_ascension)
    latitude_radians = _radians(latitude)

    # Production does not clamp; Swift `asin` of an out-of-range binary64 sum
    # yields NaN rather than raising, so mirror that instead of clamping.
    sine_altitude = (math.sin(declination) * math.sin(latitude_radians)
                     + math.cos(declination) * math.cos(latitude_radians) * math.cos(hour_angle))
    altitude = math.asin(sine_altitude) if -1.0 <= sine_altitude <= 1.0 else math.nan
    azimuth = math.atan2(
        math.sin(hour_angle),
        math.cos(hour_angle) * math.sin(latitude_radians)
        - math.tan(declination) * math.cos(latitude_radians),
    )

    return Sample(
        time=reference_seconds,
        altitude=_degrees(altitude),
        azimuth=_normalized_degrees(_degrees(azimuth) + 180),
        solar_elongation=_angular_separation((x, y, z), sun_geocentric),
    )


def observe(body: str, latitude: float, longitude: float, night_start: float, night_end: float,
            sample_interval: float = DEFAULT_SAMPLE_INTERVAL) -> list[Sample] | None:
    """`None` reproduces the production `guard end > start` bail-out."""
    start = night_start - LEAD_SECONDS
    end = night_end + TRAIL_SECONDS
    if not end > start:
        return None

    samples: list[Sample] = []
    time = start
    while time <= end:
        samples.append(position(body, latitude, longitude, time))
        time = time + sample_interval
    return samples


# --- transport ------------------------------------------------------------


def _invalid() -> ValidationError:
    return ValidationError(f"invalid {CAPABILITY_ID} input")


def _number(value: Any) -> float:
    if isinstance(value, bool) or not isinstance(value, (int, float)):
        raise _invalid()
    try:
        number = float(value)
    except OverflowError as exc:
        raise _invalid() from exc
    if not math.isfinite(number):
        raise _invalid()
    return number


def _reference_seconds(value: Any) -> float:
    if not isinstance(value, str) or not _PATTERN.fullmatch(value):
        raise _invalid()
    try:
        parsed = datetime.strptime(value, "%Y-%m-%dT%H:%M:%SZ").replace(tzinfo=timezone.utc)
    except ValueError as exc:
        raise _invalid() from exc
    seconds = (parsed - _EPOCH).total_seconds()
    if not EARLIEST_EPOCH_SECONDS <= seconds <= LATEST_EPOCH_SECONDS:
        raise _invalid()
    return seconds - _REFERENCE_EPOCH


def _timestamp(reference_seconds: float) -> str:
    if not math.isfinite(reference_seconds):
        raise _invalid()
    seconds = math.floor(reference_seconds + _REFERENCE_EPOCH)
    # isoformat preserves four-digit years even where strftime('%Y') does not.
    return (_EPOCH + timedelta(seconds=seconds)).isoformat().replace("+00:00", "Z")


def evaluate_planet_observation(input: Mapping[str, Any]) -> dict[str, Any]:
    if not isinstance(input, Mapping):
        raise _invalid()
    allowed = {"target_id", "latitude", "longitude", "night_start", "night_end",
               "sample_interval_seconds"}
    required = allowed - {"sample_interval_seconds"}
    if set(input) - allowed or not required <= set(input):
        raise _invalid()
    body = input["target_id"]
    if not isinstance(body, str) or body not in SUPPORTED_TARGET_IDS:
        raise _invalid()
    latitude = _number(input["latitude"])
    longitude = _number(input["longitude"])
    if not -90 <= latitude <= 90 or not -180 <= longitude <= 180:
        raise _invalid()
    sample_interval = (
        _number(input["sample_interval_seconds"])
        if "sample_interval_seconds" in input
        else DEFAULT_SAMPLE_INTERVAL
    )
    if not 0 < sample_interval <= MAX_SPAN_SECONDS:
        raise _invalid()
    night_start = _reference_seconds(input["night_start"])
    night_end = _reference_seconds(input["night_end"])
    if night_end - night_start > MAX_SPAN_SECONDS:
        raise _invalid()
    start = night_start - LEAD_SECONDS
    end = night_end + TRAIL_SECONDS
    if exceeds_sample_cap(start, end, sample_interval, MAX_SAMPLE_COUNT):
        raise SampleCapError(
            f"{CAPABILITY_ID} exceeds the 1.0 sample cap ({MAX_SAMPLE_COUNT} samples)"
        )
    # Integral cadence preserves strictly ordered whole-second transport and
    # avoids host-dependent repeated-addition rounding on different epochs.
    if sample_interval != math.trunc(sample_interval):
        raise _invalid()

    samples = observe(body, latitude, longitude, night_start, night_end, sample_interval)
    result: dict[str, Any] = {
        "target_id": body,
        "night_start": _timestamp(night_start),
        "night_end": _timestamp(night_end),
    }
    if samples is None:
        result["observation"] = None
        return result
    result["observation"] = {
        "sample_start": _timestamp(start),
        "sample_end": _timestamp(end),
        "samples": [
            {
                "time": _timestamp(sample.time),
                "altitude": sample.altitude,
                "azimuth": sample.azimuth,
                "solar_elongation": sample.solar_elongation,
            }
            for sample in samples
        ],
    }
    return result

File: packages/astro-engine-swift/Sources/AstroEngine/Astronomy/PlanetObservation.swift
import Foundation

/// Bodies of the production low-precision orbital-element model.
///
/// `earth` is not a target: it exists only because the model derives the
/// geocentric Sun vector from Earth's heliocentric elements, exactly as
/// production does. The user-facing recommendation candidates are
/// `recommendationCandidates`.
public enum PlanetBody: String, Sendable, Hashable, CaseIterable {
    case earth
    case venus
    case mars
    case jupiter
    case saturn

    /// Venus, Mars, Jupiter and Saturn — the production `catalog.solar_system`
    /// planet rows. Mercury, Uranus and Neptune are not production candidates
    /// and are deliberately absent from the model.
    public static let recommendationCandidates: [PlanetBody] = [.venus, .mars, .jupiter, .saturn]
}

/// One time-stamped planet position sample.
///
/// `solarElongation` is always produced by this sampler; the deterministic
/// scorer accepts a missing value because injected facts may omit it.
public struct PlanetObservationSample: Sendable, Hashable {
    public let time: Date
    /// Geometric altitude in degrees. No refraction, no parallax.
    public let altitude: Double
    /// Azimuth in degrees from north, increasing eastward, in `0..<360`.
    public let azimuth: Double
    /// Geocentric Sun–planet angular separation in degrees, in `0...180`.
    public let solarElongation: Double

    public init(time: Date, altitude: Double, azimuth: Double, solarElongation: Double) {
        self.time = time
        self.altitude = altitude
        self.azimuth = azimuth
        self.solarElongation = solarElongation
    }
}

/// Production low-precision planet astronomy, without host recommendation types.
///
/// Portable contract: contracts/procedures/planet-observation.md. This is the
/// model production ships — Schlyter's low-precision orbital elements with the
/// `JD - 2451543.5` day number, one Kepler correction term, no light-time and no
/// refraction. It is deliberately *not* an ephemeris: replacing it would move
/// every altitude and therefore every planet score.
///
/// The 900 s cadence, the two-hour lead before astronomical night, the one-hour
/// trail after it and the inclusive `while time <= end` endpoint are the
/// production quirks of `LowPrecisionPlanetAstronomyProvider` and are preserved.
public struct LowPrecisionPlanetObservationSampler: Sendable {
    /// Production cadence of `LowPrecisionPlanetAstronomyProvider`.
    public static let defaultSampleInterval: TimeInterval = 15 * 60
    /// Sampling starts this far before astronomical-night start.
    public static let leadSeconds: TimeInterval = 2 * 3600
    /// Sampling ends this far after astronomical-night end.
    public static let trailSeconds: TimeInterval = 60 * 60

    private let sampleInterval: TimeInterval

    public init(sampleInterval: TimeInterval = LowPrecisionPlanetObservationSampler.defaultSampleInterval) {
        self.sampleInterval = sampleInterval
    }

    public static func sampleStart(nightStart: Date) -> Date {
        nightStart.addingTimeInterval(-leadSeconds)
    }

    public static func sampleEnd(nightEnd: Date) -> Date {
        nightEnd.addingTimeInterval(trailSeconds)
    }

    /// `nil` reproduces the production `guard end > start` bail-out: an interval
    /// inverted by more than `leadSeconds + trailSeconds` yields no observation
    /// at all, which is not the same as an empty sample array.
    public func samples(
        body: PlanetBody,
        latitude: Double,
        longitude: Double,
        nightStart: Date,
        nightEnd: Date
    ) -> [PlanetObservationSample]? {
        let start = Self.sampleStart(nightStart: nightStart)
        let end = Self.sampleEnd(nightEnd: nightEnd)
        guard end > start else { return nil }

        var samples: [PlanetObservationSample] = []
        var time = start

        while time <= end {
            samples.append(position(
                body: body,
                latitude: latitude,
                longitude: longitude,
                at: time
            ))
            time = time.addingTimeInterval(sampleInterval)
        }

        return samples
    }

    /// Geocentric horizontal coordinates plus solar elongation at one instant.
    ///
    /// Operation ordering is load-bearing: the equatorial rotation, the
    /// `atan2` pair and the `degrees(...) + 180` azimuth normalization are
    /// reproduced from production rather than re-derived.
    public func position(
        body: PlanetBody,
        latitude: Double,
        longitude: Double,
        at date: Date
    ) -> PlanetObservationSample {
        let jd = Self.julianDate(from: date)
        // These low-precision orbital elements use Schlyter's day-number convention:
        // d = JD - 2451543.5 (rather than the J2000.0 epoch JD 2451545.0).
        let d = jd - 2_451_543.5
        let sunGeocentric = PlanetOrbitalElements.elements(for: .earth, schlyterDayNumber: d).heliocentricCoordinates()
        let planetCoordinates = PlanetOrbitalElements.elements(for: body, schlyterDayNumber: d).heliocentricCoordinates()

        let x = planetCoordinates.x + sunGeocentric.x
        let y = planetCoordinates.y + sunGeocentric.y
        let z = planetCoordinates.z + sunGeocentric.z
        let obliquity = Self.radians(23.4393 - 3.563e-7 * d)

        let equatorialX = x
        let equatorialY = y * cos(obliquity) - z * sin(obliquity)
        let equatorialZ = y * sin(obliquity) + z * cos(obliquity)
        let rightAscension = atan2(equatorialY, equatorialX)
        let declination = atan2(equatorialZ, sqrt(equatorialX * equatorialX + equatorialY * equatorialY))

        let localSiderealTime = Self.radians(Self.normalizedDegrees(
            280.460_618_37 + 360.985_647_366_29 * (jd - 2_451_545.0) + longitude
        ))
        let hourAngle = Self.normalizedRadians(localSiderealTime - rightAscension)
        let latitudeRadians = Self.radians(latitude)

        let altitude = asin(
            sin(declination) * sin(latitudeRadians)
            + cos(declination) * cos(latitudeRadians) * cos(hourAngle)
        )
        let azimuth = atan2(
            sin(hourAngle),
            cos(hourAngle) * sin(latitudeRadians) - tan(declination) * cos(latitudeRadians)
        )

        return PlanetObservationSample(
            time: date,
            altitude: Self.degrees(altitude),
            azimuth: Self.normalizedDegrees(Self.degrees(azimuth) + 180),
            solarElongation: Self.angularSeparation(
                first: (x, y, z),
                second: sunGeocentric
            )
        )
    }

    static func angularSeparation(
        first: (x: Double, y: Double, z: Double),
        second: (x: Double, y: Double, z: Double)
    ) -> Double {
        let dotProduct = first.x * second.x + first.y * second.y + first.z * second.z
        let firstMagnitude = sqrt(first.x * first.x + first.y * first.y + first.z * first.z)
        let secondMagnitude = sqrt(second.x * second.x + second.y * second.y + second.z * second.z)
        guard firstMagnitude > 0, secondMagnitude > 0 else { return 0 }

        let cosine = min(max(dotProduct / (firstMagnitude * secondMagnitude), -1), 1)
        return degrees(acos(cosine))
    }

    static func julianDate(from date: Date) -> Double {
        date.timeIntervalSince1970 / 86_400 + 2_440_587.5
    }

    static func radians(_ degrees: Double) -> Double {
        degrees * .pi / 180
    }

    static func degrees(_ radians: Double) -> Double {
        radians * 180 / .pi
    }

    static func normalizedDegrees(_ degrees: Double) -> Double {
        (degrees.truncatingRemainder(dividingBy: 360) + 360)
            .truncatingRemainder(dividingBy: 360)
    }

    static func normalizedRadians(_ radians: Double) -> Double {
        let twoPi = 2 * Double.pi
        return (radians.truncatingRemainder(dividingBy: twoPi) + twoPi)
            .truncatingRemainder(dividingBy: twoPi)
    }
}

/// Schlyter low-precision orbital elements, reproduced from production.
///
/// The eccentric-anomaly solution is the single-correction approximation
/// production ships, not an iterated Kepler solve.
struct PlanetOrbitalElements {
    let longitudeOfAscendingNode: Double
    let inclination: Double
    let argumentOfPerihelion: Double
    let semiMajorAxis: Double
    let eccentricity: Double
    let meanAnomaly: Double

    static func elements(for planet: PlanetBody, schlyterDayNumber d: Double) -> PlanetOrbitalElements {
        switch planet {
        case .earth:
            return PlanetOrbitalElements(
                longitudeOfAscendingNode: 0,
                inclination: 0,
                argumentOfPerihelion: 282.9404 + 4.70935e-5 * d,
                semiMajorAxis: 1,
                eccentricity: 0.016709 - 1.151e-9 * d,
                meanAnomaly: 356.0470 + 0.9856002585 * d
            )
        case .venus:
            return PlanetOrbitalElements(
                longitudeOfAscendingNode: 76.6799 + 2.46590e-5 * d,
                inclination: 3.3946 + 2.75e-8 * d,
                argumentOfPerihelion: 54.8910 + 1.38374e-5 * d,
                semiMajorAxis: 0.723330,
                eccentricity: 0.006773 - 1.302e-9 * d,
                meanAnomaly: 48.0052 + 1.6021302244 * d
            )
        case .mars:
            return PlanetOrbitalElements(
                longitudeOfAscendingNode: 49.5574 + 2.11081e-5 * d,
                inclination: 1.8497 - 1.78e-8 * d,
                argumentOfPerihelion: 286.5016 + 2.92961e-5 * d,
                semiMajorAxis: 1.523688,
                eccentricity: 0.093405 + 2.516e-9 * d,
                meanAnomaly: 18.6021 + 0.5240207766 * d
            )
        case .jupiter:
            return PlanetOrbitalElements(
                longitudeOfAscendingNode: 100.4542 + 2.76854e-5 * d,
                inclination: 1.3030 - 1.557e-7 * d,
                argumentOfPerihelion: 273.8777 + 1.64505e-5 * d,
                semiMajorAxis: 5.20256,
                eccentricity: 0.048498 + 4.469e-9 * d,
                meanAnomaly: 19.8950 + 0.0830853001 * d
            )
        case .saturn:
            return PlanetOrbitalElements(
                longitudeOfAscendingNode: 113.6634 + 2.38980e-5 * d,
                inclination: 2.4886 - 1.081e-7 * d,
                argumentOfPerihelion: 339.3939 + 2.97661e-5 * d,
                semiMajorAxis: 9.55475,
                eccentricity: 0.055546 - 9.499e-9 * d,
                meanAnomaly: 316.9670 + 0.0334442282 * d
            )
        }
    }

    func heliocentricCoordinates() -> (x: Double, y: Double, z: Double) {
        let meanAnomalyRadians = Self.radians(Self.normalizedDegrees(meanAnomaly))
        let eccentricAnomaly = meanAnomalyRadians
            + eccentricity * sin(meanAnomalyRadians) * (1 + eccentricity * cos(meanAnomalyRadians))

        let xv = semiMajorAxis * (cos(eccentricAnomaly) - eccentricity)
        let yv = semiMajorAxis * sqrt(1 - eccentricity * eccentricity) * sin(eccentricAnomaly)
        let trueAnomaly = atan2(yv, xv)
        let radius = sqrt(xv * xv + yv * yv)

        let node = Self.radians(longitudeOfAscendingNode)
        let inclinationRadians = Self.radians(inclination)
        let argument = trueAnomaly + Self.radians(argumentOfPerihelion)

        let x = radius * (cos(node) * cos(argument) - sin(node) * sin(argument) * cos(inclinationRadians))
        let y = radius * (sin(node) * cos(argument) + cos(node) * sin(argument) * cos(inclinationRadians))
        let z = radius * sin(argument) * sin(inclinationRadians)

        return (x, y, z)
    }

    private static func radians(_ degrees: Double) -> Double {
        degrees * .pi / 180
    }

    private static func normalizedDegrees(_ degrees: Double) -> Double {
        (degrees.truncatingRemainder(dividingBy: 360) + 360)
            .truncatingRemainder(dividingBy: 360)
    }
}
